pocs/camera/zwo.py

```diff
@@ -0,0 +1,446 @@
+        kwargs['port'] = None
+        kwargs['file_extension'] = 'fits'
+        kwargs['readout_time'] = kwargs.get('readout_time', 0.1)
+
+        # Maximum time to wait beyond exposure time for an exposure to complete
+        self._timeout = get_quantity_value(timeout, unit=u.second)
+
+        # Create a lock that wil be used to prevent overlapping exposures
+        self._exposure_lock = threading.Lock()
+        self._video_event = threading.Event()
+
+        if Camera._ASIDriver is None:
+            # Initialise the driver if it hasn't already been done
+            Camera._ASIDriver = libasi.ASIDriver(library_path=library_path)
+
+        logger = get_root_logger()
+        serial_number = kwargs.get('serial_number')
+        if not serial_number:
+            msg = "Must specify serial_number (string ID) for ZWO ASI Camera"
+            logger.error(msg)
+            raise ValueError(msg)
+
+        logger.debug("Looking for ZWO ASI camera with ID '{}'".format(serial_number))
+
+        if not Camera._ids:
+            # Not cached camera IDs, need to probe for connected cameras
+            n_cameras = Camera._ASIDriver.get_num_of_connected_cameras()
+            if n_cameras == 0:
+                msg = "No ZWO ASI camera devices found"
+                logger.error(msg)
+                warn(msg)
+                return
+
+            # Get the IDs
+            for camera_index in range(n_cameras):
+                # Can get IDs without opening cameras by parsing the name string
+                self._info = Camera._ASIDriver.get_camera_property(camera_index)
+                model, _, id = self._info['name'].partition('(')
+                if not id:
+                    logger.warning("Found ZWO ASI camera with no ID set")
+                    break
+                assert id.endswith(')'), self.logger.error("Expected ID enclosed in ()")
+                id = id[:-1]
+                Camera._ids.append(id)
+
+            logger.debug('Connected ASI ZWO cameras: {}'.format(Camera._ids))
+
+        try:
+            self._camera_index = Camera._ids.index(serial_number)
+        except ValueError:
+            msg = "Could not find ZWO ASI camera with ID '{}'".format(serial_number)
+            logger.error(msg)
+            warn(msg)
+            return
+        else:
+            logger.debug("Found ZWO ASI Camera with ID '{}' at index {}".format(
+                serial_number, self._camera_index))
+
+        if serial_number in Camera._assigned_ids:
+            msg = "ZWO ASI Camera with ID '{}' already in use".format(serial_number)
+            logger.error(msg)
+            warn(msg)
+            return
+
+        super().__init__(name, *args, **kwargs)
+        self.connect()
+        assert self.is_connected, error.PanError("Could not connect to {}".format(self))
+
+        Camera._assigned_ids.append(self.uid)
+
+        self._setter_try('ccd_set_point', 'set_point', set_point)
```

Comment:
I see `_setter_try` used 3 times. In which of these would you be expecting the set to fail and if you know that in advance why bother calling it?

Reply:
Hmm, it is of marginal usefulness. All it really does is convert `NotImplementedError` exceptions to warnings in the case where the user tried to set `set_point`, `gain` or `image_type` when initialising a camera the doesn't support setting those.  Our expert users probably _should_ have a fairly good idea of the cameras that they're configuring, so throwing a fatal `NotImplementedError` if they make a mistake in `pocs.yaml` is perhaps reasonable.